src/ppl/src/IOPlacer.cpp

```diff
@@ -1284,6 +1284,90 @@ void IOPlacer::findSlotsForTopLayer()
         top_layer_slots_.push_back({false, false, Point(x, y), top_grid_.layer, Edge::invalid});
       }
     }
+
+    filterObstructedSlotsForTopLayer();
+  }
+}
+
+void IOPlacer::filterObstructedSlotsForTopLayer()
+{
+  // Collect top_grid_ obstructions
+  std::vector<std::unique_ptr<odb::Rect>> obstructions;
+
+  // Get routing obstructions
+  for (odb::dbObstruction* obstruction : block_->getObstructions()) {
+    odb::dbBox* box = obstruction->getBBox();
+    if (box->getTechLayer()->getRoutingLevel() == top_grid_.layer) {
+      obstructions.push_back(std::make_unique<odb::Rect>());
```

Comment:
It seems like it would be more readable to follow this order instead.

```c++
std::make_unique<odb::Rect> obstruction_rect = std::make_unique<odb::Rect>();
box->getBox(*obstructions.back());
obstructions.push_back(obstruction_rect);
```

Reply:
good point, I'll modify accordingly

Reply:
@QuantamHD I ended up dropping the unique_ptr, the Rect is so small anyways copying it shouldn't negatively impact performance.